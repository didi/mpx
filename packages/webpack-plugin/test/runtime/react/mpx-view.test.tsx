import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react-native'
import MpxView from '../../../lib/runtime/components/react/mpx-view'
import MpxInlineText from '../../../lib/runtime/components/react/mpx-inline-text'

// Mock mpx-portal
jest.mock('../../../lib/runtime/components/react/mpx-portal', () => {
  const mockReact = require('react')
  return {
    __esModule: true,
    default: mockReact.forwardRef((props: any, ref: any) => {
      return mockReact.createElement('View', { ...props, ref })
    })
  }
})

describe('MpxView with MpxInlineText', () => {
  it('should render basic view', () => {
    const { toJSON } = render(
      <MpxView testID="basic-view">
        <MpxInlineText>Basic View Content</MpxInlineText>
      </MpxView>
    )
    
    expect(screen.getByTestId('basic-view')).toBeTruthy()
    expect(screen.getByText('Basic View Content')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should apply custom styles', () => {
    const customStyle = { 
      backgroundColor: '#f0f0f0',
      padding: 20,
      flex: 1
    }
    
    const { toJSON } = render(
      <MpxView style={customStyle} testID="styled-view">
        <MpxInlineText>Styled Content</MpxInlineText>
      </MpxView>
    )
    
    const viewElement = screen.getByTestId('styled-view')
    expect(viewElement).toBeTruthy()
    expect(viewElement.props.style).toMatchObject(customStyle)
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle nested views', () => {
    const { toJSON } = render(
      <MpxView testID="parent-view">
        <MpxView testID="child-view-1">
          <MpxInlineText>Child 1</MpxInlineText>
        </MpxView>
        <MpxView testID="child-view-2">
          <MpxInlineText>Child 2</MpxInlineText>
        </MpxView>
      </MpxView>
    )
    
    expect(screen.getByTestId('parent-view')).toBeTruthy()
    expect(screen.getByTestId('child-view-1')).toBeTruthy()
    expect(screen.getByTestId('child-view-2')).toBeTruthy()
    expect(screen.getByText('Child 1')).toBeTruthy()
    expect(screen.getByText('Child 2')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle accessibility props', () => {
    render(
      <MpxView 
        testID="accessible-view"
        accessibilityLabel="Main container"
        accessibilityHint="Contains important content"
        accessibilityRole="main"
      >
        <MpxInlineText>Accessible Content</MpxInlineText>
      </MpxView>
    )
    
    const viewElement = screen.getByTestId('accessible-view')
    expect(viewElement).toBeTruthy()
    expect(viewElement.props.accessibilityLabel).toBe('Main container')
    expect(viewElement.props.accessibilityHint).toBe('Contains important content')
    expect(viewElement.props.accessibilityRole).toBe('main')
  })

  it('should handle complex nested structure', () => {
    const { toJSON } = render(
      <MpxView testID="complex-structure" style={{ flex: 1 }}>
        <MpxView testID="header" style={{ height: 50 }}>
          <MpxInlineText>Header Content</MpxInlineText>
        </MpxView>
        <MpxView testID="body" style={{ flex: 1 }}>
          <MpxView testID="section-1">
            <MpxInlineText>Section 1</MpxInlineText>
          </MpxView>
          <MpxView testID="section-2">
            <MpxInlineText>Section 2</MpxInlineText>
          </MpxView>
        </MpxView>
        <MpxView testID="footer" style={{ height: 50 }}>
          <MpxInlineText>Footer Content</MpxInlineText>
        </MpxView>
      </MpxView>
    )
    
    // 验证所有组件都能被找到
    expect(screen.getByTestId('complex-structure')).toBeTruthy()
    expect(screen.getByTestId('header')).toBeTruthy()
    expect(screen.getByTestId('body')).toBeTruthy()
    expect(screen.getByTestId('section-1')).toBeTruthy()
    expect(screen.getByTestId('section-2')).toBeTruthy()
    expect(screen.getByTestId('footer')).toBeTruthy()
    
    // 验证文本内容
    expect(screen.getByText('Header Content')).toBeTruthy()
    expect(screen.getByText('Section 1')).toBeTruthy()
    expect(screen.getByText('Section 2')).toBeTruthy()
    expect(screen.getByText('Footer Content')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle empty view', () => {
    const { toJSON } = render(
      <MpxView testID="empty-view" />
    )
    
    const viewElement = screen.getByTestId('empty-view')
    expect(viewElement).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle view with single child', () => {
    const { toJSON } = render(
      <MpxView testID="single-child-view">
        <MpxInlineText>Single Child</MpxInlineText>
      </MpxView>
    )
    
    const viewElement = screen.getByTestId('single-child-view')
    expect(viewElement).toBeTruthy()
    expect(screen.getByText('Single Child')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should handle view with multiple text children', () => {
    const { toJSON } = render(
      <MpxView testID="multiple-text-view">
        <MpxInlineText>First Text</MpxInlineText>
        <MpxInlineText>Second Text</MpxInlineText>
        <MpxInlineText>Third Text</MpxInlineText>
      </MpxView>
    )
    
    const viewElement = screen.getByTestId('multiple-text-view')
    expect(viewElement).toBeTruthy()
    expect(screen.getByText('First Text')).toBeTruthy()
    expect(screen.getByText('Second Text')).toBeTruthy()
    expect(screen.getByText('Third Text')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  // Basic View functionality tests
  describe('Basic View functionality', () => {
    it('should handle MPX touch events', () => {
      const bindtap = jest.fn()
      const bindtouchstart = jest.fn()
      const bindtouchend = jest.fn()
      render(
        <MpxView 
          testID="touchable-view" 
          bindtap={bindtap}
          bindtouchstart={bindtouchstart}
          bindtouchend={bindtouchend}
        >
          <MpxInlineText>Touch me</MpxInlineText>
        </MpxView>
      )
      
      // 验证 MPX 事件属性被正确接收
      const viewElement = screen.getByTestId('touchable-view')
      expect(viewElement).toBeTruthy()
      
      // 验证组件确实接收了 MPX 事件绑定属性
      // 注意：由于 useInnerProps 的复杂性，我们主要验证组件能正确接收这些属性
      expect(bindtap).toBeDefined()
      expect(bindtouchstart).toBeDefined()
      expect(bindtouchend).toBeDefined()
    })

    it('should handle background image properties', () => {
      const backgroundStyle = {
        backgroundColor: '#ff0000'
      }
      
      const { toJSON } = render(
        <MpxView 
          testID="background-image-view"
          enable-background={true}
          style={backgroundStyle}
        >
          <MpxInlineText>Background Image</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('background-image-view')
      expect(viewElement).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle linear gradient background', () => {
      const gradientStyle = {
        backgroundImage: 'linear-gradient(45deg, #ff0000 0%, #00ff00 50%, #0000ff 100%)',
        backgroundSize: ['100%', '100%']
      }
      
      const { toJSON } = render(
        <MpxView 
          testID="gradient-view"
          enable-background={true}
          style={gradientStyle}
        >
          <MpxInlineText>Gradient Background</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('gradient-view')
      expect(viewElement).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle complex background properties', () => {
      const complexBackgroundStyle = {
        backgroundColor: '#f0f0f0',
        borderRadius: 10
      }
      
      const { toJSON } = render(
        <MpxView 
          testID="complex-background-view"
          enable-background={true}
          enable-fast-image={true}
          style={complexBackgroundStyle}
        >
          <MpxInlineText>Complex Background</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('complex-background-view')
      expect(viewElement).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle flex layout', () => {
      const { toJSON } = render(
        <MpxView 
          testID="flex-view"
          style={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'space-between',
            alignItems: 'center'
          }}
        >
          <MpxInlineText>Item 1</MpxInlineText>
          <MpxInlineText>Item 2</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('flex-view')
      expect(viewElement).toBeTruthy()
      expect(viewElement.props.style.display).toBe('flex')
      expect(viewElement.props.style.flexDirection).toBe('row')
      expect(toJSON()).toMatchSnapshot()
    })

    it('should properly forward refs', () => {
      const ref = React.createRef()
      render(
        <MpxView ref={ref} testID="ref-view">
          <MpxInlineText>Ref View</MpxInlineText>
        </MpxView>
      )
      
      expect(ref.current).toBeTruthy()
    })
  })

  // Edge cases
  describe('Edge cases', () => {
    it('should handle undefined style', () => {
      const { toJSON } = render(
        <MpxView testID="undefined-style-view" style={undefined}>
          <MpxInlineText>Undefined Style</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('undefined-style-view')
      expect(viewElement).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle null children', () => {
      const { toJSON } = render(
        <MpxView testID="null-children-view">
          {null}
          <MpxInlineText>Valid Child</MpxInlineText>
          {null}
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('null-children-view')
      expect(viewElement).toBeTruthy()
      expect(screen.getByText('Valid Child')).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle mixed children types', () => {
      const { toJSON } = render(
        <MpxView testID="mixed-children-view">
          <MpxInlineText>Text Child</MpxInlineText>
          {42}
          <MpxView>
            <MpxInlineText>Nested View</MpxInlineText>
          </MpxView>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('mixed-children-view')
      expect(viewElement).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    it('should handle boolean props correctly', () => {
      render(
        <MpxView 
          testID="boolean-props-view"
          pointerEvents="none"
          removeClippedSubviews={true}
          collapsable={false}
        >
          <MpxInlineText>Boolean Props</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('boolean-props-view')
      expect(viewElement).toBeTruthy()
      expect(viewElement.props.pointerEvents).toBe('none')
      expect(viewElement.props.removeClippedSubviews).toBe(true)
      expect(viewElement.props.collapsable).toBe(false)
    })

    it('should handle zero and negative values in styles', () => {
      const styleWithZeroAndNegative = {
        margin: 0,
        padding: -5,
        width: 0,
        height: -10
      }
      
      render(
        <MpxView testID="zero-negative-view" style={styleWithZeroAndNegative}>
          <MpxInlineText>Zero Negative</MpxInlineText>
        </MpxView>
      )
      
      const viewElement = screen.getByTestId('zero-negative-view')
      expect(viewElement).toBeTruthy()
      expect(viewElement.props.style).toMatchObject(styleWithZeroAndNegative)
    })
  })
})