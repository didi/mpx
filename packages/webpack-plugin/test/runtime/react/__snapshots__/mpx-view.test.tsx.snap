// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MpxView with MpxInlineText Basic View functionality should handle background image properties 1`] = `
<View
  style={
    {
      "backgroundColor": "#ff0000",
    }
  }
  testID="background-image-view"
>
  <Text
    allowFontScaling={false}
  >
    Background Image
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText Basic View functionality should handle complex background properties 1`] = `
<View
  enable-fast-image={true}
  style={
    {
      "backgroundColor": "#f0f0f0",
      "borderRadius": 10,
    }
  }
  testID="complex-background-view"
>
  <Text
    allowFontScaling={false}
  >
    Complex Background
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText Basic View functionality should handle flex layout 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexBasis": "auto",
      "flexDirection": "row",
      "flexShrink": 1,
      "flexWrap": "nowrap",
      "justifyContent": "space-between",
    }
  }
  testID="flex-view"
>
  <Text
    allowFontScaling={false}
  >
    Item 1
  </Text>
  <Text
    allowFontScaling={false}
  >
    Item 2
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText Basic View functionality should handle linear gradient background 1`] = `
<View
  style={{}}
  testID="gradient-view"
>
  <View
    onLayout={[Function]}
    style={
      {
        "bottom": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <Text
    allowFontScaling={false}
  >
    Gradient Background
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText Edge cases should handle mixed children types 1`] = `
<View
  style={{}}
  testID="mixed-children-view"
>
  <Text
    allowFontScaling={false}
  >
    Text Child
  </Text>
  42
  <View
    style={{}}
  >
    <Text
      allowFontScaling={false}
    >
      Nested View
    </Text>
  </View>
</View>
`;

exports[`MpxView with MpxInlineText Edge cases should handle null children 1`] = `
<View
  style={{}}
  testID="null-children-view"
>
  <Text
    allowFontScaling={false}
  >
    Valid Child
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText Edge cases should handle undefined style 1`] = `
<View
  style={{}}
  testID="undefined-style-view"
>
  <Text
    allowFontScaling={false}
  >
    Undefined Style
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText should apply custom styles 1`] = `
<View
  style={
    {
      "backgroundColor": "#f0f0f0",
      "flex": 1,
      "padding": 20,
    }
  }
  testID="styled-view"
>
  <Text
    allowFontScaling={false}
  >
    Styled Content
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText should handle complex nested structure 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
  testID="complex-structure"
>
  <View
    style={
      {
        "height": 50,
      }
    }
    testID="header"
  >
    <Text
      allowFontScaling={false}
    >
      Header Content
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
    testID="body"
  >
    <View
      style={{}}
      testID="section-1"
    >
      <Text
        allowFontScaling={false}
      >
        Section 1
      </Text>
    </View>
    <View
      style={{}}
      testID="section-2"
    >
      <Text
        allowFontScaling={false}
      >
        Section 2
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 50,
      }
    }
    testID="footer"
  >
    <Text
      allowFontScaling={false}
    >
      Footer Content
    </Text>
  </View>
</View>
`;

exports[`MpxView with MpxInlineText should handle empty view 1`] = `
<View
  style={{}}
  testID="empty-view"
/>
`;

exports[`MpxView with MpxInlineText should handle nested views 1`] = `
<View
  style={{}}
  testID="parent-view"
>
  <View
    style={{}}
    testID="child-view-1"
  >
    <Text
      allowFontScaling={false}
    >
      Child 1
    </Text>
  </View>
  <View
    style={{}}
    testID="child-view-2"
  >
    <Text
      allowFontScaling={false}
    >
      Child 2
    </Text>
  </View>
</View>
`;

exports[`MpxView with MpxInlineText should handle view with multiple text children 1`] = `
<View
  style={{}}
  testID="multiple-text-view"
>
  <Text
    allowFontScaling={false}
  >
    First Text
  </Text>
  <Text
    allowFontScaling={false}
  >
    Second Text
  </Text>
  <Text
    allowFontScaling={false}
  >
    Third Text
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText should handle view with single child 1`] = `
<View
  style={{}}
  testID="single-child-view"
>
  <Text
    allowFontScaling={false}
  >
    Single Child
  </Text>
</View>
`;

exports[`MpxView with MpxInlineText should render basic view 1`] = `
<View
  style={{}}
  testID="basic-view"
>
  <Text
    allowFontScaling={false}
  >
    Basic View Content
  </Text>
</View>
`;
